### PR TITLE
Another round of docstring updates

### PR DESCRIPTION
**Pull request by tonypnode**
I'm expecting conflicts on this one, but I will work through them. I will also go back and add the defaults. I didn't get a chance to go back and do it this round. Let me know where I need to adjust.

### CODE REVIEW

**Comment by tonypnode** [comment]
Travis is failing, but it looks like a problem with the build setup. I didn't change any code.. just docstrings haha.

**Comment by tonypnode** [comment]
ahh.. I see what the deal is... it's checking for pep8 and my strings are too long. Fixing now.

File: netmiko/base_connection.py [hunk 1]
```diff
@@ -519,6 +621,11 @@ def establish_connection(self, width=None, height=None):
         Authentication failure will generate a NetMikoAuthenticationException
 
         width and height are needed for Fortinet paging setting.
+
+        :param width: Specified width of the terminal window for VT100
+        :type width int
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: missing colon
```py
"""
:type width: int
"""
```

File: netmiko/base_connection.py [hunk 2]
```diff
@@ -160,7 +179,18 @@ def __enter__(self):
...
+        """Gracefully close connection on Context Manager exit.
+
+        :param exc_type:
+        :type exc_type:
+
+        :param exc_value:
+        :type exc_value:
+
+        :param traceback:
+        :type traceback:
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Missing descriptions

File: netmiko/base_connection.py [hunk 3]
```diff
@@ -294,6 +341,15 @@ def _read_channel_expect(self, pattern='', re_flags=0, max_loops=None):
 
         There are dependencies here like determining whether in config_mode that are actually
         depending on reading beyond pattern.
+
+        :param pattern: The pattern used to identify device prompt
+        :type pattern: str
+
+        :param re_flags: Regex flags used in conjunction with pattern to search for prompt
+        :type re_flags: re module flags
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Flags are integers, please change to `int`.

File: netmiko/base_connection.py [hunk 4]
```diff
@@ -368,7 +430,14 @@ def read_until_pattern(self, *args, **kwargs):
...
 
     def read_until_prompt_or_pattern(self, pattern='', re_flags=0):
-        """Read until either self.base_prompt or pattern is detected. Return ALL data available."""
+        """Read until either self.base_prompt or pattern is detected. Return ALL data available.
+
+        :param pattern: The pattern used to identify device prompt
+        :type pattern: str
+
+        :param re_flags: Regex flags used in conjunction with pattern to search for prompt
+        :type re_flags: re module flags
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Change to `int`, it's important for static type checkers.

File: netmiko/base_connection.py [hunk 5]
```diff
@@ -559,10 +666,28 @@ def establish_connection(self, width=None, height=None):
...
 
     def _test_channel_read(self, count=40, pattern=""):
-        """Try to read the channel (generally post login) verify you receive data back."""
+        """Try to read the channel (generally post login) verify you receive data back.
+
+        :param count: the number of times to check the channel for data
+        :type count: int
+
+        :param pattern: Signifying the device prompt has returned and to break out of the loop
+        :type: str
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Missing variable name `:type pattern: str`

File: netmiko/base_connection.py [hunk 6]
```diff
@@ -559,10 +666,28 @@ def establish_connection(self, width=None, height=None):
...
+        """
 
         def _increment_delay(main_delay, increment=1.1, maximum=8):
-            """Increment sleep time to a maximum value."""
+            """Increment sleep time to a maximum value.
+
+            :param main_delay: Pri sleep factor for data to return from the channel
+            :type int
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Missing argument name. `:type main_delay: int`

File: netmiko/base_connection.py [hunk 7]
```diff
@@ -559,10 +666,28 @@ def establish_connection(self, width=None, height=None):
...
+
+            :param increment: Sec sleep factor for waiting for data to return from channel
+            :type increment: float
+
+            :param maximum: Max delay to sleep when waiting for data to return from the channel
+            :type int
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Missing argument name. `:type maximum: int`

File: netmiko/base_connection.py [hunk 8]
```diff
@@ -559,10 +666,28 @@ def establish_connection(self, width=None, height=None):
...
+            :
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Remove this colon please

File: netmiko/base_connection.py [hunk 9]
```diff
@@ -941,6 +1173,9 @@ def send_config_from_file(self, config_file=None, **kwargs):
         The file is processed line-by-line and each command is sent down the
         SSH channel.
 
+        :param config_file: Configuration file to be sent do the device
+        :type:
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Missing argument name and description

File: netmiko/base_connection.py [hunk 10]
```diff
@@ -955,6 +1190,24 @@ def send_config_set(self, config_commands=None, exit_config_mode=True, delay_fac
         The commands will be executed one after the other.
 
         Automatically exits/enters configuration mode.
+
+        :param config_commands: Multiple commands to be sent to the device
+        :type config_commands: iterable of strings
```
Comment (xavierhardy) [inline netmiko/base_connection.py]: Nesting types using `of` is supported by AFAIK at least PyCharm. `list of str` would be valid for instance and is correctly parsed by the PyCharm static type checker.

**Comment by xavierhardy** [comment]
You need to pull the changes from ktbyers's repository and rebase on his `master` branch.

https://git-scm.com/book/en/v2/Git-Branching-Rebasing

**Comment by tonypnode** [comment]
not sure how this is going to go. My major conflicts were due to not rebasing in a while, amongst some other rookie mistakes. I am going to push again now and see how it goes. If it's all jacked up, I'll just canx this PR and push another.

**Comment by ktbyers** [comment]
Okay, just let me know if you have issues. I will try to get to this merged in the next couple of weeks.

**Comment by tonypnode** [comment]
looks like everything took. I'll start working on one of the other files. If you have anything specific you need, let me know

**Comment by ktbyers** [comment]
This was merged here:

https://github.com/ktbyers/netmiko/commit/682bfc26c30ae4f9a05ed079c036c51ad0369fa9


There were some changes to it in the commit right after it.

**Pull request closed without merging**